File: src/leetcode/2025/12/Solution205.cpp
#include <string>
#include <unordered_map>
using namespace std;

class Solution205
{
public:
    bool isIsomorphic(string s, string t)
    {
        if (s.length() != t.length())
        {
            return false;
        }

        int mapS[256] = {0};
        int mapT[256] = {0};

        for (size_t i = 0; i < s.length(); i++)
        {
            char charS = s[i];
            char charT = t[i];

            if (mapS[charS] != mapT[charT])
            {
                return false;
            }

            mapS[charS] = i + 1;
            mapT[charT] = i + 1;
        }

        return true;
    }
};

int main()
{
    Solution205 solution;
    string s1 = "egg";
    string t1 = "add";
    bool result1 = solution.isIsomorphic(s1, t1);
    printf("%s\n", result1 ? "true" : "false");
    // Expected output: true

    string s2 = "foo";
    string t2 = "bar";
    bool result2 = solution.isIsomorphic(s2, t2);
    printf("%s\n", result2 ? "true" : "false");
    // Expected output: false

    string s3 = "paper";
    string t3 = "title";
    bool result3 = solution.isIsomorphic(s3, t3);
    printf("%s\n", result3 ? "true" : "false");
    // Expected output: true

    return 0;
}

File: src/leetcode/2025/12/Solution205.java
public class Solution205 {
    public boolean isIsomorphic(String s, String t) {
        if (s.length() != t.length()) {
            return false;
        }

        int[] mapS = new int[256];
        int[] mapT = new int[256];

        for (int i = 0; i < s.length(); i++) {
            char charS = s.charAt(i);
            char charT = t.charAt(i);

            if (mapS[charS] != mapT[charT]) {
                return false;
            }

            mapS[charS] = i + 1;
            mapT[charT] = i + 1;
        }

        return true;
    }

    public static void main(String[] args) {
        Solution205 solution = new Solution205();

        // Example test cases
        String s1 = "egg";
        String t1 = "add";
        System.out.println(solution.isIsomorphic(s1, t1)); // true

        String s2 = "foo";
        String t2 = "bar";
        System.out.println(solution.isIsomorphic(s2, t2)); // false

        String s3 = "paper";
        String t3 = "title";
        System.out.println(solution.isIsomorphic(s3, t3)); // true
    }
}
